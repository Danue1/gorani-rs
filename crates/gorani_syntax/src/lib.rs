pub type SyntaxNode = rowan::SyntaxNode<SyntaxKind>;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum SyntaxKind {
    ERROR,

    /// https://spec.graphql.org/draft/#sec-Document
    DOCUMENT,

    /// https://spec.graphql.org/draft/#ExecutableDocument
    EXECUTABLE_DOCUMENT,

    /// https://spec.graphql.org/draft/#TypeSystemDocument
    TYPE_SYSTEM_DOCUMENT,

    /// https://spec.graphql.org/draft/#TypeSystemExtensionDocument
    TYPE_SYSTEM_EXTENSION_DOCUMENT,

    /// https://spec.graphql.org/draft/#Definition
    DEFINITION,

    /// https://spec.graphql.org/draft/#ExecutableDefinition
    EXECUTABLE_DEFINITION,

    /// https://spec.graphql.org/draft/#SchemaDefinition
    SCHEMA_DEFINITION,

    /// https://spec.graphql.org/draft/#TypeSystemDefinition
    TYPE_SYSTEM_DEFINITION,

    /// https://spec.graphql.org/draft/#TypeSystemDefinitionOrExtension
    TYPE_SYSTEM_DEFINITION_OR_EXTENSION,

    /// https://spec.graphql.org/draft/#OperationDefinition
    OPERATION_DEFINITION,

    /// https://spec.graphql.org/draft/#FragmentDefinition
    FRAGMENT_DEFINITION,

    /// https://spec.graphql.org/draft/#VariablesDefinition
    VARIABLES_DEFINITION,

    /// https://spec.graphql.org/draft/#VariableDefinition
    VARIABLE_DEFINITION,

    /// https://spec.graphql.org/draft/#RootOperationTypeDefinition
    ROOT_OPERATION_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#DirectiveDefinition
    DIRECTIVE_DEFINITION,

    /// https://spec.graphql.org/draft/#ArgumentsDefinition
    ARGUMENTS_DEFINITION,

    /// https://spec.graphql.org/draft/#InputValueDefinition
    INPUT_VALUE_DEFINITION,

    /// https://spec.graphql.org/draft/#InputFieldsDefinition
    INPUT_FIELDS_DEFINITION,

    /// https://spec.graphql.org/draft/#FieldsDefinition
    FIELDS_DEFINITION,

    /// https://spec.graphql.org/draft/#FieldDefinition
    FIELD_DEFINITION,

    /// https://spec.graphql.org/draft/#EnumValuesDefinition
    ENUM_VALUES_DEFINITION,

    /// https://spec.graphql.org/draft/#EnumValueDefinition
    ENUM_VALUE_DEFINITION,

    /// https://spec.graphql.org/draft/#TypeDefinition
    TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#ScalarTypeDefinition
    SCALAR_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#ObjectTypeDefinition
    OBJECT_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#InterfaceTypeDefinition
    INTERFACE_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#UnionTypeDefinition
    UNION_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#EnumTypeDefinition
    ENUM_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#InputObjectTypeDefinition
    INPUT_OBJECT_TYPE_DEFINITION,

    /// https://spec.graphql.org/draft/#TypeSystemExtension
    TYPE_SYSTEM_EXTENSION,

    /// https://spec.graphql.org/draft/#SchemaExtension
    SCHEMA_EXTENSION,

    /// https://spec.graphql.org/draft/#TypeExtension
    TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#ScalarTypeExtension
    SCALAR_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#ObjectTypeExtension
    OBJECT_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#InterfaceTypeExtension
    INTERFACE_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#UnionTypeExtension
    UNION_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#EnumTypeExtension
    ENUM_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#InputObjectTypeExtension
    INPUT_OBJECT_TYPE_EXTENSION,

    /// https://spec.graphql.org/draft/#UnionMemberTypes
    UNION_MEMBER_TYPES,

    /// https://spec.graphql.org/draft/#ImplementsInterfaces
    IMPLEMENTS_INTERFACES,

    /// https://spec.graphql.org/draft/#OperationType
    OPERATION_TYPE,

    /// https://spec.graphql.org/draft/#DirectiveLocations
    DIRECTIVE_LOCATIONS,

    /// https://spec.graphql.org/draft/#DirectiveLocation
    DIRECTIVE_LOCATION,

    /// https://spec.graphql.org/draft/#ExecutableDirectiveLocation
    EXECUTABLE_DIRECTIVE_LOCATION,

    /// https://spec.graphql.org/draft/#TypeSystemDirectiveLocation
    TYPE_SYSTEM_DIRECTIVE_LOCATION,

    /// https://spec.graphql.org/draft/#Description
    DESCRIPTION,

    /// https://spec.graphql.org/draft/#Variable
    VARIABLE,

    /// https://spec.graphql.org/draft/#DefaultValue
    DEFAULT_VALUE,

    /// https://spec.graphql.org/draft/#Directives
    DIRECTIVES,

    /// https://spec.graphql.org/draft/#Directive
    DIRECTIVE,

    /// https://spec.graphql.org/draft/#Arguments
    ARGUMENTS,

    /// https://spec.graphql.org/draft/#Argument
    ARGUMENT,

    /// https://spec.graphql.org/draft/#Type
    TYPE,

    /// https://spec.graphql.org/draft/#NamedType
    NAMED_TYPE,

    /// https://spec.graphql.org/draft/#ListType
    LIST_TYPE,

    /// https://spec.graphql.org/draft/#NonNullType
    NON_NULL_TYPE,

    /// https://spec.graphql.org/draft/#SelectionSet
    SELECTION_SET,

    /// https://spec.graphql.org/draft/#Selection
    SELECTION,

    /// https://spec.graphql.org/draft/#Field
    FIELD,

    /// https://spec.graphql.org/draft/#Alias
    ALIAS,

    /// https://spec.graphql.org/draft/#FragmentSpread
    FRAGMENT_SPREAD,

    /// https://spec.graphql.org/draft/#FragmentName
    FRAGMENT_NAME,

    /// https://spec.graphql.org/draft/#sec-Type-Conditions
    TYPE_CONDITION,

    /// https://spec.graphql.org/draft/#sec-Inline-Fragments
    INLINE_FRAGMENT,

    /// https://spec.graphql.org/draft/#Value
    VALUE,

    /// https://spec.graphql.org/draft/#IntValue
    INT_VALUE,

    /// https://spec.graphql.org/draft/#FloatValue
    FLOAT_VALUE,

    /// https://spec.graphql.org/draft/#StringValue
    STRING_VALUE,

    /// https://spec.graphql.org/draft/#BooleanValue
    BOOLEAN_VALUE,

    /// https://spec.graphql.org/draft/#NullValue
    NULL_VALUE,

    /// https://spec.graphql.org/draft/#EnumValue
    ENUM_VALUE,

    /// https://spec.graphql.org/draft/#ListValue
    LIST_VALUE,

    /// https://spec.graphql.org/draft/#ObjectValue
    OBJECT_VALUE,

    /// https://spec.graphql.org/draft/#ObjectField
    OBJECT_FIELD,

    /// https://spec.graphql.org/draft/#BlockString
    BLOCK_STRING,

    /// https://spec.graphql.org/draft/#IntegerPart
    INTEGER_PART,

    /// https://spec.graphql.org/draft/#NegativeSign
    NEGATIVE_SIGN,

    /// https://spec.graphql.org/draft/#NonZeroDigit
    NON_ZERO_DIGIT,

    /// https://spec.graphql.org/draft/#ExponentIndicator
    EXPONENT_INDICATOR,

    /// https://spec.graphql.org/draft/#FractionalPart
    FRACTIONAL_PART,

    /// https://spec.graphql.org/draft/#ExponentPart
    EXPONENT_PART,

    /// https://spec.graphql.org/draft/#Sign
    SIGN,

    /// https://spec.graphql.org/draft/#sec-White-Space
    WHITE_SPACE,

    /// https://spec.graphql.org/draft/#sec-Line-Terminators
    LINE_TERMINATOR,

    /// https://spec.graphql.org/draft/#sec-Comments
    COMMENT,

    /// https://spec.graphql.org/draft/#sec-Insignificant-Commas
    INSIGNIFICANT_COMMA,

    /// https://spec.graphql.org/draft/#sec-Language.Source-Text.Ignored-Tokens
    IGNORED_TOKEN,

    /// https://spec.graphql.org/draft/#Punctuator
    PUNCTUATOR,

    /// https://spec.graphql.org/draft/#sec-Names
    NAME,

    KEYWORD_FRAGMENT,     // fragment
    KEYWORD_QUERY,        // query
    KEYWORD_MUTATION,     // mutation
    KEYWORD_SUBSCRIPTION, // subscription
    KEYWORD_TRUE,         // true
    KEYWORD_FALSE,        // false
    KEYWORD_NULL,         // null
    KEYWORD_REPEATABLE,   // repeatable
    KEYWORD_ON,           // on
    KEYWORD_SCALAR,       // scalar
    KEYWORD_SCHEMA,       // schema
    KEYWORD_TYPE,         // type
    KEYWORD_INTERFACE,    // interface
    KEYWORD_UNION,        // union
    KEYWORD_ENUM,         // enum
    KEYWORD_INPUT,        // input
    KEYWORD_DIRECTIVE,    // directive
    KEYWORD_IMPLEMENTS,   // implements
    KEYWORD_EXTEND,       // extend
    KEYWORD_EXTENDS,      // extends

    SYMBOL_LEFT_PARENS,   // (
    SYMBOL_RIGHT_PARENS,  // )
    SYMBOL_COLON,         // :
    SYMBOL_EQUAL,         // =
    SYMBOL_AT,            // @
    SYMBOL_DOLLAR,        // $
    SYMBOL_AMPERSAND,     // &
    SYMBOL_LEFT_BRACKET,  // [
    SYMBOL_RIGHT_BRACKET, // ]
    SYMBOL_LEFT_BRACE,    // {
    SYMBOL_RIGHT_BRACE,   // }
    SYMBOL_EXCLAMATION,   // !
    SYMBOL_PIPE,          // |
    SYMBOL_SPREAD,        // ...

    // https://spec.graphql.org/draft/#sec-Location
    LOCATION_QUERY,               // QUERY
    LOCATION_MUTATION,            // MUTATION
    LOCATION_SUBSCRIPTION,        // SUBSCRIPTION
    LOCATION_FIELD,               // FIELD
    LOCATION_FRAGMENT_DEFINITION, // FRAGMENT_DEFINITION
    LOCATION_FRAGMENT_SPREAD,     // FRAGMENT_SPREAD
    LOCATION_INLINE_FRAGMENT,     // INLINE_FRAGMENT
    LOCATION_VARIABLE_DEFINITION, // VARIABLE_DEFINITION

    // https://spec.graphql.org/draft/#sec-Location
    LOCATION_SCHEMA,                 // SCHEMA
    LOCATION_SCALAR,                 // SCALAR
    LOCATION_OBJECT,                 // OBJECT
    LOCATION_FIELD_DEFINITION,       // FIELD_DEFINITION
    LOCATION_ARGUMENT_DEFINITION,    // ARGUMENT_DEFINITION
    LOCATION_INTERFACE,              // INTERFACE
    LOCATION_UNION,                  // UNION
    LOCATION_ENUM,                   // ENUM
    LOCATION_ENUM_VALUE,             // ENUM_VALUE
    LOCATION_INPUT_OBJECT,           // INPUT_OBJECT
    LOCATION_INPUT_FIELD_DEFINITION, // INPUT_FIELD_DEFINITION
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}

impl rowan::Language for SyntaxKind {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        unsafe { std::mem::transmute(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}
