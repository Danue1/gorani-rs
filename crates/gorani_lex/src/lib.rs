use gorani_syntax::SyntaxKind;
use std::borrow::Cow;

pub fn lex(mut source: &str) -> impl Iterator<Item = (SyntaxKind, usize, Cow<str>)> {
    std::iter::from_fn(move || {
        let c = source.chars().next()?;
        let (kind, len, token) = lex_token(c, source);
        let token = token
            .map(Cow::Owned)
            .unwrap_or_else(|| Cow::Borrowed(&source[..len]));
        source = &source[len..];

        Some((kind, len, token))
    })
}

fn lex_token(c: char, source: &str) -> (SyntaxKind, usize, Option<String>) {
    match c {
        '(' => (SyntaxKind::SYMBOL_LEFT_PARENS, 1, None),
        ')' => (SyntaxKind::SYMBOL_RIGHT_PARENS, 1, None),
        ':' => (SyntaxKind::SYMBOL_COLON, 1, None),
        '=' => (SyntaxKind::SYMBOL_EQUAL, 1, None),
        '@' => (SyntaxKind::SYMBOL_AT, 1, None),
        '$' => (SyntaxKind::SYMBOL_DOLLAR, 1, None),
        '&' => (SyntaxKind::SYMBOL_AMPERSAND, 1, None),
        '[' => (SyntaxKind::SYMBOL_LEFT_BRACKET, 1, None),
        ']' => (SyntaxKind::SYMBOL_RIGHT_BRACKET, 1, None),
        '{' => (SyntaxKind::SYMBOL_LEFT_BRACE, 1, None),
        '}' => (SyntaxKind::SYMBOL_RIGHT_BRACE, 1, None),
        '!' => (SyntaxKind::SYMBOL_EXCLAMATION, 1, None),
        '|' => (SyntaxKind::SYMBOL_PIPE, 1, None),
        '.' => {
            if source.starts_with("...") {
                (SyntaxKind::SYMBOL_SPREAD, 3, None)
            } else {
                (SyntaxKind::ERROR, 1, None)
            }
        }
        'a'..='z' | 'A'..='Z' | '_' => {
            let len = source
                .chars()
                .take_while(|c| c.is_ascii_alphanumeric() || *c == '_')
                .count();

            (SyntaxKind::NAME, len, None)
        }
        '-' => {
            let (kind, len) = number(true, source);

            (kind, len, None)
        }
        '0'..='9' => {
            let (kind, len) = number(false, source);

            (kind, len, None)
        }
        '"' => {
            let (kind, len, token) = if source.starts_with("\"\"\"") {
                block_string(source)
            } else {
                return single_line_string(source);
            };

            (kind, len, Some(token))
        }
        ',' | ' ' | '\t' | '\n' | '\r' => {
            let len = source
                .chars()
                .take_while(|c| matches!(c, ',' | ' ' | '\t' | '\n' | '\r'))
                .count();

            (SyntaxKind::IGNORED_TOKEN, len, None)
        }
        '#' => {
            let line = match source.lines().next() {
                Some(line) => line.len(),
                None => source.len(),
            };

            (SyntaxKind::COMMENT, line, None)
        }
        _ => (SyntaxKind::ERROR, 1, None),
    }
}

fn block_string(_source: &str) -> (SyntaxKind, usize, String) {
    std::todo!();
}

fn single_line_string(source: &str) -> (SyntaxKind, usize, Option<String>) {
    let rest = &source[1..];
    let Some(line) = rest.lines().next() else {
        return (SyntaxKind::ERROR, source.len(), None);
    };

    let mut len = 1;
    let mut string = String::new();
    let mut rest = line.chars();
    while let Some(c) = rest.next() {
        len += 1;

        match c {
            '"' => break,
            '\\' => match rest.next() {
                Some('"') => {
                    len += 1;
                    string.push('"');
                }
                Some('\\') => {
                    len += 1;
                    string.push('\\');
                }
                Some('/') => {
                    len += 1;
                    string.push('/');
                }
                Some('b') => {
                    len += 1;
                    string.push('\x08');
                }
                Some('f') => {
                    len += 1;
                    string.push('\x0C');
                }
                Some('n') => {
                    len += 1;
                    string.push('\n');
                }
                Some('r') => {
                    len += 1;
                    string.push('\r');
                }
                Some('t') => {
                    len += 1;
                    string.push('\t');
                }
                Some('u') => std::todo!(),
                _ => return (SyntaxKind::ERROR, line.len() + 1, None),
            },
            _ => string.push(c),
        }
    }

    (SyntaxKind::STRING_VALUE, len, Some(string))
}

fn number(negate: bool, source: &str) -> (SyntaxKind, usize) {
    let (negate, rest) = if negate {
        (1, &source[1..])
    } else {
        (0, source)
    };
    let (integer_part, rest) = if rest.starts_with('0') {
        (1, &rest[1..])
    } else {
        let len = rest.chars().take_while(|c| c.is_ascii_digit()).count();
        if len == 0 {
            return (SyntaxKind::ERROR, negate);
        }

        (len, &rest[len..])
    };
    let (fraction_part, rest) = if rest.starts_with('.') {
        let len = rest
            .chars()
            .skip(1)
            .take_while(|c| c.is_ascii_digit())
            .count();
        if len == 0 {
            let len = integer_part + negate;

            return (SyntaxKind::INT_VALUE, len);
        }

        (len + 1, &rest[len + 1..])
    } else {
        (0, rest)
    };
    let exponent_part = if rest.starts_with(|c| matches!(c, 'e' | 'E')) {
        let rest = &rest[1..];
        let (sign_part, rest) = if rest.starts_with(|c| matches!(c, '+' | '-')) {
            (1, &rest[1..])
        } else {
            (0, rest)
        };
        let len = rest.chars().take_while(|c| c.is_ascii_digit()).count();
        if len == 0 {
            let len = integer_part + fraction_part + negate;

            return (SyntaxKind::INT_VALUE, len);
        }

        sign_part + len + 1
    } else {
        0
    };

    if fraction_part == 0 && exponent_part == 0 {
        let len = integer_part + negate;

        (SyntaxKind::INT_VALUE, len)
    } else {
        let len = integer_part + fraction_part + exponent_part + negate;

        (SyntaxKind::FLOAT_VALUE, len)
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn punctuators() {
        let tokens = super::lex(r#"():=@$&[]{}!|..."#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn names() {
        let tokens = super::lex(r#"_ _123 foo foo123 true false null"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn integers() {
        let tokens = super::lex(r#"0 123 -0 -123"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn no_integers() {
        let tokens = super::lex(r#"0x 0. 0e 123_456 -123_456"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn floats_with_fraction() {
        let tokens = super::lex(r#"0.0 123.0 123.456 -0.0 -123.0 -123.456"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn floats_with_exponent() {
        let tokens = super::lex(r#"0e0 123e0 123e456 -0e0 -123e0 -123e456"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn floats_with_fraction_and_exponent() {
        let tokens = super::lex(
            r#"0.0e0 123.0e0 123.456e0 123.0e456 123.456e789 -0.0e0 -123.0e0 -123.456e0 -123.0e456 -123.456e789"#,
        );
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn single_line_string() {
        let tokens = super::lex(r#""" "foo" "foo\"bar" "foo\nbar" "foo\rbar" "foo\r\nbar""#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[ignore]
    #[test]
    fn single_line_string_with_escape_unicode() {
        let tokens = super::lex(r#""\u1234""#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[ignore]
    #[test]
    fn block_string() {
        let tokens = super::lex(r#"""""" """"""""#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn comments() {
        let tokens = super::lex(r#"# foo"#);
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn ignored_tokens() {
        let tokens = super::lex(" ,\t\n\r");
        let tokens: Vec<_> = tokens.collect();
        insta::assert_debug_snapshot!(tokens);
    }
}
