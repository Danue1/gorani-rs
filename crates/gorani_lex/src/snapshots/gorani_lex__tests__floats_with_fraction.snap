---
source: crates/gorani_lex/src/lib.rs
expression: tokens
---
[
    (
        FLOAT_VALUE,
        3,
        "0.0",
    ),
    (
        IGNORED_TOKEN,
        1,
        " ",
    ),
    (
        FLOAT_VALUE,
        5,
        "123.0",
    ),
    (
        IGNORED_TOKEN,
        1,
        " ",
    ),
    (
        FLOAT_VALUE,
        7,
        "123.456",
    ),
    (
        IGNORED_TOKEN,
        1,
        " ",
    ),
    (
        FLOAT_VALUE,
        4,
        "-0.0",
    ),
    (
        IGNORED_TOKEN,
        1,
        " ",
    ),
    (
        FLOAT_VALUE,
        6,
        "-123.0",
    ),
    (
        IGNORED_TOKEN,
        1,
        " ",
    ),
    (
        FLOAT_VALUE,
        8,
        "-123.456",
    ),
]
